### ISSUE TITLE
Cursor moves back to beginning of line when exiting :FZF popup on Vim 9 (after #209)

### ISSUE THREAD

**Issue by cxw42** [editorconfig/editorconfig-vim#224]
> Hi, recently I have an issue with editorconfig-vim after this PR merged.

`test.vimrc`:
``` vim
set nocompatible

call plug#begin()
Plug 'editorconfig/editorconfig-vim'
Plug 'junegunn/fzf', { 'dir': '~/.fzf', 'do': './install --all' }
call plug#end()

autocmd FileType c EditorConfigReload
```

`.editorconfig`:
```
root = true

[*]
tab_width = 8
```

`test.c`:
``` c
#include <stdio.h>

int main() {
  int x = 1;
  printf("%d\n", x);
  return 0;
}
```

How to reproduce the problem:
- Run `vim -u test.vimrc test.c`
- Place the cursor anywhere except the first column
- Run `:FZF`
- Press <key>Esc</key> to close the popup
- The cursor goes to the first column of line 1

I couldn't reduce `test.vimrc` further as I don't know much about popup windows. I've also run `git bisect`, and it pointed e269673.

_Originally posted by @yous in https://github.com/editorconfig/editorconfig-vim/issues/209#issuecomment-1680605824_

**Comment by cxw42** [comment]
@yous thanks for reporting!  What is the _expected_ behaviour?  What did it used to do that it doesn't any more?

Could you provide me the command-line or script you used for `git bisect`?  Or did you manually test and use `good`/`bad`?

Much appreciated!

**Comment by yous** [comment]
The expected behavior is, after I press Esc, the cursor should go back to the previous location. But currently, the cursor goes to the first line after the popup window closed, so screen scrolls to the top.

I ran `git bisect good` / `git bisect bad` manually.

**Comment by cxw42** [comment]
Thanks for the details!

**Comment by midgleyc** [comment]
I also have this problem. As a tiny bit more information, it happens on any pop-up specified by `fzf` (e.g. `:W`, `:Files`), and it goes to the top of the file when the pop-up opens, before you press escape to close. The cursor didn't used to move at all.

**Comment by cxw42** [comment]
@yous @midgleyc which Vim and FZF versions?  I haven't been able to repro yet.  (But I'm using `set rtp+=` rather than vim-plug, so I need to try that next.)
```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Oct 16 2023 18:14:13)
Included patches: 1-213, 1840, 214-579, 1969, 580-1854, 1857, 1855-1857, 1331, 1858, 1858-1859, 1873, 1860-1969, 1992, 1970-1992, 2010, 1993-2269, 3612, 3625, 3669, 3741, 1847
```

**Comment by yous** [comment]
I can reproduce in WSL with these versions:

```console
$ vim --version
VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Nov 11 2023 17:59:33)
Included patches: 1-2100
```

```console
$ fzf --version
0.44.1 (d7d2ac3)
```

**Comment by midgleyc** [comment]
Also WSL, with:
```
$ vim --version
VIM - Vi IMproved 9.0 (2022 Jun 28, compiled May 10 2022 08:40:37)
Included patches: 1-749
```
```
$ fzf --version
0.29 (devel)
```

**Comment by cxw42** [comment]
Thanks for the quick responses!  I'll try Vim 9.

**Comment by cxw42** [comment]
I am able to repro on Vim 9 --- the behaviour is different at e26967348^ than at that commit.
```
VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Nov 19 2023 11:11:34)
Included patches: 1-2100

$ fzf --version
0.24.0-1 (3304f28)

fzf vim plugin 0.44.1
```

**Additionally,** I see the same difference at 345a5a32a808.  I think I must have accidentally hit rebase-merge on #209 --- sorry!  e26967348 is part of the feature branch for #209, and is not in the history of `master`.

**Comment by cxw42** [comment]
@yous @midgleyc It looks like this might be a Vim bug.  Would you be willing to try my MCVE at <https://github.com/cxw42/vim9-tabstop-repro-2023-11> and see if you observe the same issue?  Instructions in the README.  Once I hear back from you, I will report on the Vim issue tracker.  Thanks!

Summary: It appears `setbufvar(&tabstop)` changes the cursor position when used in `BufFilePost`.

Some options include:
- Revert 345a5a3 (but this might cause old issues related to handling of no-name buffers to resurface)
- Revert 345a5a3 and remove `BufNew` --- that is the case for which I made the `%`/`<abuf>` change, as far as I can recall.
- In `BufFilePost` (or other autocmds?), don't apply settings if we already did so and the buffer's name is unchanged

**Comment by chrisbra** [comment]
what is the purpose of the `BufFilePost` ? It triggers when renaming the terminal buffer of the FZF buffer to a terminal buffer for `!sh`. The terminal seems to be running in the same window as test.c at least temporarily and so does the `setbufvar()` option trigger for the terminal window, Vim tries to validate the cursor position of the first line (which is empty) and therefor positions the cursor on the start of the line in that window.

**Comment by chrisbra** [comment]
I'd suggest to add a condition like this to it to fix it:

```vim
function! s:SetTabstop(which) abort
    let l:tabstop = 8
    let l:bufnr = str2nr(expand('<abuf>'))
    " skip this event for terminal buffers
    if bufname(l:bufnr) =~ '^!\w*sh$'
      return
    endif
```

**Comment by yous** [comment]
I can confirm that observed behaviors mentioned in https://github.com/cxw42/vim9-tabstop-repro-2023-11 are reproducible in macOS.

```console
$ vim --version
VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Nov 11 2023 17:59:33)
macOS version - arm64
Included patches: 1-2100
```

```console
$ fzf --version
0.44.1 (d7d2ac3)
```

Also when I added `if bufname(l:bufnr) =~ ...` condition that @chrisbra mentioned, cursor doesn't move.

**Comment by cxw42** [comment]
@chrisbra @yous thanks very much!

> The terminal seems to be running in the same window as test.c at least temporarily

This seems like either a vim or an fzf bug --- @chrisbra do you have any thoughts on which it might be?

**Comment by cxw42** [comment]
> what is the purpose of the `BufFilePost`?

It was added in aadfe238e3beb342c80e46eb2e928c7141a762c3 "Reload config on file name changes".  It looks from `:help autocmd.txt` like `BufFilePost` is the only way to catch `:file` and `:saveas`, so we need to keep it.

**Comment by chrisbra** [comment]
Not sure, It may be the normal way how the terminal is started inside the popup. I would rather add the condition to skip the auto command for buffer names starting with ! (indicating a terminal). Or perhaps that auto command should use <buffer> and only trigger for buffers where editorconfig is actually active, instead of for all buffers.

Am 24. November 2023, um 20:37, schrieb Chris White ***@***.***>:

@chrisbra @yous thanks very much!

The terminal seems to be running in the same window as test.c at least temporarily

This seems like either a vim or an fzf bug --- @chrisbra do you have any thoughts on which it might be?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.￼Message ID: ***@***.***>

[ { ***@***.***": "<a href="http://schema.org">http://schema.org</a>", ***@***.***": "EmailMessage", "potentialAction": { ***@***.***": "ViewAction", "target": "<a href="https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826033736">https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826033736</a>", "url": "<a href="https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826033736">https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826033736</a>", "name": "View Issue" }, "description": "View this Issue on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", "url": "<a href="https://github.com">https://github.com</a>" } } ]

**Comment by cxw42** [comment]
@chrisbra why does 
```
call setbufvar(l:abuf, '&tabstop', l:tabstop)
```
behave differently from
```
" bufnr() == l:abuf
let &l:tabstop = l:tabstop
```
?  (`:help eval.txt` says that the latter corresponds to `:setlocal`, so I also tried that, and it does not repro.)  I don't understand why "Vim tries to validate the cursor position" when using `setbufvar` but not when using `let &l:`/`:setlocal`.  Thanks in advance for any information you can provide!

**Comment by chrisbra** [comment]
One is switching buffers, before setting the option value,  the other is just setting the buffer local variable for the current buffer.

Am 24. November 2023, um 21:06, schrieb Chris White ***@***.***>:

@chrisbra why does

call setbufvar(l:abuf, '&tabstop', l:tabstop) 

behave differently from

" bufnr() == l:abuf let &l:tabstop = l:tabstop 

? (:help eval.txt says that the latter corresponds to :setlocal, so I also tried that, and it does not repro.) I don't understand why "Vim tries to validate the cursor position" when using setbufvar but not when using let &l:/:setlocal. Thanks in advance for any information you can provide!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.￼Message ID: ***@***.***>

[ { ***@***.***": "<a href="http://schema.org">http://schema.org</a>", ***@***.***": "EmailMessage", "potentialAction": { ***@***.***": "ViewAction", "target": "<a href="https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826051347">https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826051347</a>", "url": "<a href="https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826051347">https://github.com/editorconfig/editorconfig-vim/issues/224#issuecomment-1826051347</a>", "name": "View Issue" }, "description": "View this Issue on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", "url": "<a href="https://github.com">https://github.com</a>" } } ]

**Comment by cxw42** [comment]
@chrisbra thanks for the info!

@yous @midgleyc could you please try the code in #229 and see if it resolves the issue?  It works for me, but I would appreciate confirmation before merging.  Thank you!

**Comment by yous** [comment]
I found another problem with #229,

.editorconfig:
```
root = true

[*]
tab_width = 8

[*.c]
indent_style = space
indent_size = 2
```

test.c:
``` c
#include <stdio.h>

int main() {
}
```

test.vimrc:
``` vim
set nocompatible

call plug#begin()
Plug 'editorconfig/editorconfig-vim'
call plug#end()

autocmd FileType c EditorConfigReload
```

1. Run `vim -u test.vimrc test.c`
2. See `set sw`
3. Vim shows `shiftwidth=8`, not `shiftwidth=2`

**Comment by yous** [comment]
Ah, `l:bufnr` should be `a:bufnr` in `bufname(l:bufnr) !~# '^!\w*sh$'`.

**Comment by misraelson** [comment]
Just wondering if someone can post the fix here for folks. Currently setting up a new laptop with VIM9 and experiencing this problem. Should I paste this into my `.vimrc` ?
```
function! s:SetTabstop(which) abort
    let l:tabstop = 8
    let l:bufnr = str2nr(expand('<abuf>'))
    " skip this event for terminal buffers
    if bufname(l:bufnr) =~ '^!\w*sh$'
      return
    endif
```

**Comment by cxw42** [comment]
@misraelson If you clone this repo and add it to the beginning of your runtimepath, I think it will take precedence and you will get the fix.  I'm not sure, though, as I haven't migrated to vim9 yet.  Any vim9 folks know for sure?

I can tell you that pasting `s:SetTabstop` into your .vimrc won't fix it, since `s:` means "script", i.e., the current file.  You could probably copy the current contents of this repo over the version of the plugin that ships with vim 9, though.
